frame/staking/src/ledger.rs

```diff
@@ -0,0 +1,258 @@
+use {
+	codec::{Decode, Encode, MaxEncodedLen},
+	scale_info::TypeInfo,
+	sp_runtime::traits::Zero,
+};
+
+impl<T: Config> StakingLedger<T> {
+	#[cfg(any(feature = "runtime-benchmarks", test))]
+	pub fn default_from(stash: T::AccountId) -> Self {
+		Self {
+			stash,
+			total: Zero::zero(),
+			active: Zero::zero(),
+			unlocking: Default::default(),
+			claimed_rewards: Default::default(),
+			controller: Default::default(),
+		}
+	}
+
+	/// Returns a new instance of a staking ledger.
+	///
+	/// The [`Ledger`] storage is not mutated. In order to store, `StakingLedger::update` must be
+	/// called on the returned staking ledger.
+	///
+	/// Note: as the controller accounts are being deprecated, the stash account is the same as the
+	/// controller account.
+	pub fn new(
+		stash: T::AccountId,
+		active_stake: BalanceOf<T>,
+		total_stake: BalanceOf<T>,
+		unlocking: BoundedVec<UnlockChunk<BalanceOf<T>>, T::MaxUnlockingChunks>,
+		claimed_rewards: BoundedVec<EraIndex, T::HistoryDepth>,
+	) -> Self {
+		Self {
+			stash: stash.clone(),
+			active: active_stake,
+			total: total_stake,
+			unlocking,
+			claimed_rewards,
+			// controllers are deprecated and mapped 1-1 to stashes.
+			controller: Some(stash),
+		}
+	}
+
+	/// Returns the paired account, if any.
+	///
+	/// A "pair" refers to the tuple (stash, controller). If the input is a
+	/// [`StakingAccount::Stash`] variant, its pair account will be of type
+	/// [`StakingAccount::Controller`] and vice-versa.
+	///
+	/// This method is meant to abstract from the runtime development the difference between stash
+	/// and controller. This will be deprecated once the controller is fully deprecated as well.
+	pub(crate) fn paired_account(account: StakingAccount<T::AccountId>) -> Option<T::AccountId> {
+		match account {
+			StakingAccount::Stash(stash) => <Bonded<T>>::get(stash),
+			StakingAccount::Controller(controller) =>
+				<Ledger<T>>::get(&controller).map(|ledger| ledger.stash),
+		}
+	}
+
+	/// Returns whether a given account is bonded.
+	pub(crate) fn is_bonded(account: StakingAccount<T::AccountId>) -> bool {
+		match account {
+			StakingAccount::Stash(stash) => <Bonded<T>>::contains_key(stash),
+			StakingAccount::Controller(controller) => <Ledger<T>>::contains_key(controller),
+		}
+	}
+
+	/// Returns a staking ledger, if it is bonded and it exists in storage.
+	///
+	/// This getter can be called with either a controller or stash account, provided that the
+	/// account is properly wrapped in the respective [`StakingAccount`] variant. This is meant to
+	/// abstract the concept of controller/stash accounts from the caller.
+	pub(crate) fn get(account: StakingAccount<T::AccountId>) -> Result<StakingLedger<T>, Error<T>> {
+		let controller = match account {
+			StakingAccount::Stash(stash) => <Bonded<T>>::get(stash).ok_or(Error::<T>::NotStash),
+			StakingAccount::Controller(controller) => Ok(controller),
+		}?;
+
+		<Ledger<T>>::get(&controller)
```

Comment:
I am not suggesting you do this as it is too much in the fringes, but hear me out: 

```rust
// NewType for stash
pub struct Stash<T>(T::AccountId)
// NewType for controller 
pub struct Controller<T>(T::AccountId)

enum StakingAccount {
	Stash(Stash),
	Controller(Controller)
}
```

Now you can actually make the key of `Ledger` be `Controller`, not an opaque `AccountId`.

If deemed worthy, could be a nice follow-up, but I am worried it would have too much clutter, and once Controller is gone, it is all in vein. 

But I hope the idea is clear. We are pury complicating things so that we can use Rust's type system for more safety.